backend/service/auditsink/slack/slack.go

```diff
@@ -104,3 +123,90 @@ func formatText(username string, event *auditv1.RequestEvent) string {
 
 	return messageText
 }
+
+// FormatCustomText parses out the audit event metdata request for the custom slack message text
+func FormatCustomText(message string, event *auditv1.RequestEvent) (string, error) {
+	tmpl, err := template.New("customText").Funcs(funcMap).Parse(message)
+	if err != nil {
+		return "", err
+	}
+
+	data, err := getAuditMetadata(event)
+	if err != nil {
+		return "", err
+	}
+
+	var buf bytes.Buffer
+	if err := tmpl.Execute(&buf, data); err != nil {
+		return "", err
+	}
+
+	return buf.String(), nil
+}
+
+// returns the API request/response details saved in an audit event
+func getAuditMetadata(event *auditv1.RequestEvent) (*auditEventMetadata, error) {
+	var requestMetadata map[string]interface{}
+	var responseMetadata map[string]interface{}
+
+	// proto -> json
+	reqJSON, err := protojson.Marshal(event.RequestMetadata.Body)
+	if err != nil {
+		return nil, err
+	}
+	respJSON, err := protojson.Marshal(event.ResponseMetadata.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	// json -> map[string]interface{}
+	err = json.Unmarshal(reqJSON, &requestMetadata)
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(respJSON, &responseMetadata)
+	if err != nil {
+		return nil, err
+	}
+
+	return &auditEventMetadata{
+		Request:  requestMetadata,
+		Response: responseMetadata,
+	}, nil
+}
+
+// for inputs that are type slice/map, returns a formatted slack list OR `N/A` if the slice/map is empty
+func slackList(data interface{}) (string, error) {
+	var text string
+	const listItemFormat = "\n- %v"
+	const mapItemFormat = "\n- %v: %v"
+
+	switch reflect.TypeOf(data).Kind() {
+	case reflect.Slice:
+		s := reflect.ValueOf(data)
+		if s.Len() == 0 {
+			return "`N/A`", nil
```

Comment:
for cases when a slice or map is empty (indicating that nothing was updated/removed etc), have been debating on what's better:
"Removed Annotations":
"Removed Annotations": `N/A`

Alternatively in the Go template itself you can write if/else and use `len` to specify unique defaults but trying to cut down on how much templating we do in the config

Reply:
`None` as a single item in a list is a nice option here.

For the template use case, you could add a helper function which replaces any empty list with a slice containing `["None"]`, e.g. `{{ .MyVar | noneIfEmpty }}`

Reply:
updated to None & wrote a TODO to add more helper funcs